File: leetcode/java/lengthOfLongestSubstr.java
/*
 * given a string of lower characters, find the length of longest sub string without repeat character
 * */
import java.util.*;
import java.io.*;

public class lengthOfLongestSubstr{
    private int min(int a, int b){
        return a<b ? a : b;
    }
    public int lengthoflongestsubstr(String s){
        if(s.isEmpty())        return 0;
        int[] pos = new int[26];
        for(int i=0;i<26;i++){
            pos[i] = -1;
        }
        pos[s.charAt(0) - 'a'] = 0;
        int[] length = new int[s.length()];
        length[0] = 1;
        int res=1;
        for(int i=1;i<s.length();i++){
            char ch = s.charAt(i);
            length[i] = min(length[i-1] + 1, i - pos[ch - 'a']);
            pos[ch - 'a'] = i;
            if(length[i] > res)    res = length[i];
        }
        return res;
    }

    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);
        lengthOfLongestSubstr ls = new lengthOfLongestSubstr();
        while(true){
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            System.out.println(ls.lengthoflongestsubstr(str));
        }
        return;
    }
}

File: leetcode/src/lengthOfLongestSubstr.cpp
/*
 * given a string of lower case, find the length of longest sub string without duplicate character(s)
 *
 * */

#include <string>
#include <iostream>
#include <vector>

using namespace std;

class Solution{

public:
    /*
     * once scan, in time O(n) and space O(n). smart solution
     * */
    int lengthOfLongestSubstr(const string& str){
        if(str.empty())    return 0;

        const int n = str.size();
        int res = 1;

        int pos[26];    // last position of char [a-z] appeared in s
        for(int i = 0; i < 26; ++i){
            pos[i] = -1;    // -1 as invalid position, it tells no duplicate char
        }
        pos[str[0] - 'a'] = 0;

        vector<int> lengths(n, 0); // longest substr without any duplicate char ending at [i]
        lengths[0] = 1;

        for(int i = 1; i < n; i++){
            lengths[i] = min(lengths[i-1] + 1, i - pos[str[i] - 'a']);    // optimal substructure of DP
            pos[str[i] - 'a'] = i;    // update last occurance of str[i]
            res = max(res, lengths[i]);
        }

        return res;
    }
};

/* unit test is in ../cpp_unittest/lengthOfLongestSubstr_unittest */
